Civi/Financeextras/Event/Subscriber/CreditNoteInvoiceSubscriber.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+namespace Civi\Financeextras\Event\Subscriber;
+
+use DateTime;
+use Civi\Financeextras\Event\CreditNoteMailedEvent;
+use Civi\Financeextras\Event\CreditNoteDownloadedEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class CreditNoteInvoiceSubscriber implements EventSubscriberInterface {
+
+  /**
+   * {@inheritDoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      CreditNoteMailedEvent::NAME => 'createMailActivity',
+      CreditNoteDownloadedEvent::NAME => 'createDownloadActivity',
+    ];
+  }
+
+  /**
+   * Create activity for the creditnote invoice download.
+   *
+   * @param \Civi\Financeextras\Event\CreditNoteDownloadedEvent $e
+   *   The registration event. Add new tokens using register().
+   */
+  public function createDownloadActivity(CreditNoteDownloadedEvent $e) {
+    $activityType = 'Downloaded Invoice';
+    $subject = 'Downloaded Credit Note PDF';
+    $targetContactId = $this->getCreditNoteContactId($e->getCreditNoteId());
+    $attachment = $this->storeFile($e->getCreditNoteInvoice()['html'], $e->getCreditNoteInvoice()['format']);
+    $this->createActivity([$targetContactId], $activityType, $subject, $attachment);
+  }
+
+  /**
+   * Create activity for the creditnote invoice mail.
+   *
+   * @param \Civi\Financeextras\Event\CreditNoteMailedEvent $e
+   *   The registration event. Add new tokens using register().
+   */
+  public function createMailActivity(CreditNoteMailedEvent $e) {
+    $activityType = 'Emailed Invoice';
+    $attachment = $this->storeFile($e->getCreditNoteInvoice()['html'], $e->getCreditNoteInvoice()['format']);
+    $this->createActivity($e->getMailedContacts(), $activityType, $e->getSubject(), $attachment);
+  }
+
+  /**
+   * Create Credit note invoice activity
+   *
+   * @param array $targetContactIds
+   * @param string $type
+   * @param string $subject
+   * @param array $attachment
+   */
+  private function createActivity($targetContactIds, $type, $subject, $attachment) {
+    $now = (new DateTime())->format('YmdHis');
+    $currentUser = \CRM_Core_Session::singleton()->get('userID');
+    \Civi\Api4\Activity::create()
+      ->addValue('subject', $subject)
+      ->addValue('target_contact_id', $targetContactIds)
```

Comment:
@olayiwola-compucorp  Just to confirm that we can pass arrays of target contact IDs here?

Reply:
Yes, confirmed
<img width="472" alt="Screenshot 2023-06-28 at 10 04 37" src="https://github.com/compucorp/io.compuco.financeextras/assets/85277674/f8ea1a22-d87d-4745-b513-70bda2a490b2">